File: binary_tree/max_path_sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import math
from typing import Optional

from binary_tree.binary_tree import TreeNode

"""
https://leetcode.com/problems/binary-tree-maximum-path-sum/
Time: O(N)
Space: O(N)
"""
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.maxSum = -math.inf
        
        def helper(node):
            if not node:
                return 0
            
            max_left_sum = max(helper(node.left), 0)
            max_right_sum = max(helper(node.right), 0)
            
            current_sum = node.val + max_left_sum + max_right_sum
            
            self.maxSum = max(self.maxSum, current_sum)
            
            return node.val + max(max_left_sum, max_right_sum)
        
        helper(root)
        
        return self.maxSum

File: binary_tree/max_path_sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 * https://leetcode.com/problems/binary-tree-maximum-path-sum
 * Time: O(N)
 * Space: O(N)
 */
class Solution {
    int maxSum = Integer.MIN_VALUE;
    
    public int maxPathSum(TreeNode root) {
     
        this.maxPathSumRecursive(root);
        
        return this.maxSum;
    }
    
    private int maxPathSumRecursive(TreeNode node){
        
        if(node == null){
            return 0;
        }
        
        int leftMax = Math.max(maxPathSumRecursive(node.left), 0);
        int rightMax = Math.max(maxPathSumRecursive(node.right), 0);
        
        this.maxSum = Math.max(this.maxSum, leftMax + rightMax + node.val);
        
        return Math.max(leftMax, rightMax) + node.val;
    }
}